From 55bfae5c7d3b1e09f36750a94a73c97d14bf201d Mon Sep 17 00:00:00 2001
From: Leo Shen <i@szclsya.me>
Date: Tue, 21 Apr 2026 17:00:42 -0400
Subject: [PATCH 1/6] qnx: remove features not used by QNX

- /usr merge, no longer using /bin, /sbin, /lib, /usr/sbin and /usr/libexec
  - also applies to openrc
- add qnx target triplets
- remove -pyc auto splitting
- remove auto so dependency, only prompt a warning
- remove capability detections
- remove nftables auto splitting
- remove multi-libc support
Upstream-Status: QNX-specific (from qnx-ports)

---
 abuild-sudo.c   |  10 +--
 abuild.in       | 193 +++++++++++++++++++++++++-----------------------
 functions.sh.in |  62 +---------------
 3 files changed, 107 insertions(+), 158 deletions(-)

diff --git a/abuild-sudo.c b/abuild-sudo.c
index df3a38090c..ec55d3eff1 100644
--- a/abuild-sudo.c
+++ b/abuild-sudo.c
@@ -23,11 +23,11 @@
 #endif
 
 static const char* valid_cmds[] = {
-	"/bin/adduser",
-	"/usr/sbin/adduser",
-	"/bin/addgroup",
-	"/usr/sbin/addgroup",
-	"/sbin/apk",
+	"/usr/bin/adduser",
+	"/usr/bin/adduser",
+	"/usr/bin/addgroup",
+	"/usr/bin/addgroup",
+	"/usr/bin/apk",
 	"/usr/bin/abuild-rmtemp",
 	NULL
 };
diff --git a/abuild.in b/abuild.in
index 1dfcb1fe3f..a5efb5e8b6 100644
--- a/abuild.in
+++ b/abuild.in
@@ -1,4 +1,4 @@
-#!/bin/ash -e
+#!/bin/bash -e
 # vim: set filetype=sh:
 
 # abuild - build apk packages (light version of makepkg)
@@ -491,12 +491,12 @@ initdcheck() {
 		esac
 
 		case "$line" in
-		*sbin/openrc-run)
+		*usr/bin/openrc-run)
 			;;
-		*sbin/runscript)
-			warning "$i is not an openrc #!/sbin/openrc-run"
+		*usr/bin/runscript)
+			warning "$i is not an openrc #!/usr/bin/openrc-run"
 			;;
-		*)	error "$i is not an openrc #!/sbin/openrc-run"
+		*)	error "$i is not an openrc #!/usr/bin/openrc-run"
 			return 1
 			;;
 		esac
@@ -594,7 +594,7 @@ subpkg_set() {
 	subpkgarch=${_splitarch#*:}
 	if [ "$subpkgarch" = "$_splitarch" -o -z "$subpkgarch" ]; then
 		case "$subpkgname" in
-		*-doc | *-openrc | *-lang | *sh-completion | *-pyc | *-nftrules | *-systemd | *-udev) subpkgarch="noarch" ;;
+		*-doc | *-openrc | *-lang | *sh-completion) subpkgarch="noarch" ;;
 		*) subpkgarch="$pkgarch" ;;
 		esac
 	elif [ "$subpkgarch" = "all" ]; then
@@ -804,7 +804,7 @@ postcheck() {
 
 	# /usr merge
 	if ! options_has "!usrmerge"; then
-		for i in bin sbin lib; do
+		for i in bin sbin lib usr/sbin usr/libexec; do
 			if [ -L "$dir"/$i ] || [ -e "$dir"/$i ]; then
 				warning "Packages must not put anything under /$i, use /usr/${i##*s} instead"
 			fi
@@ -816,12 +816,6 @@ postcheck() {
 		find "$dir" -name '*.la' -type f -delete
 	fi
 
-	# look for /usr/lib/charset.alias
-	if [ -e "$dir"/usr/lib/charset.alias ] \
-			&& ! options_has "charset.alias"; then
-		error "Found /usr/lib/charset.alias"
-		e=1
-	fi
 	# look for /etc/init.d and /etc/conf.d
 	if [ -e "$dir"/etc/init.d -o -e "$dir"/etc/conf.d ] \
 			&& ! is_openrc_pkg "$name"; then
@@ -860,13 +854,6 @@ postcheck() {
 			e=1
 		fi
 	fi
-	# look for pycache
-	# wildcard should always get the system python dir, and this is faster than
-	# trying to calculate the python version.
-	i="$(find "$dir"/usr/lib/python* \( -type d -a -name "__pycache__" \) 2>/dev/null || :)"
-	if [ -n "$i" ] && [ "${name%-pyc}" = "$name" ]; then
-		warning "Found __pycache__ but package name doesn't end with -pyc"
-	fi
 
 	# check that we don't have any files names with newline
 	i=$(find "$dir" -name $'*\n*')
@@ -899,20 +886,6 @@ postcheck() {
 			e=1
 		fi
 	fi
-	# test capabilities on executables
-	# see: https://gitlab.alpinelinux.org/alpine/tsc/-/issues/45
-	getcap -r "$dir" | (local r=true execothers; while read -r line; do
-		local filename="${line% *}"
-		if ! options_has "setcap"; then
-			error "Found binary with extra capabilities: $filename"
-			r=false
-		fi
-
-		execothers="$(find "$filename" -perm -o+x)"
-		if [ -n "$execothers" ]; then
-			warning "Found setcap binary executable by others: $filename"
-		fi
-	done; $r) || e=1
 
 	# test for textrels
 	if ! options_has "textrels"; then
@@ -1467,7 +1440,7 @@ real_so_path() {
 		[ -e "$1"/$so ] && realpath "$1/$so" && return 0
 		shift
 	done
-	error "$so: path not found"
+	error "$so: shared library not found"
 	return 1
 }
 
@@ -1483,13 +1456,52 @@ find_so_files() {
 	return 0
 }
 
+subpkg_provides_prefixed_so() {
+	if [ -n "$sonameprefix" ] && grep -q -w "^$sonameprefix$1" \
+		"$pkgbasedir"/.control.*/.provides-so 2>/dev/null; then
+		grep -l -H -w "^$sonameprefix$1" \
+			"$pkgbasedir"/.control.*/.provides-so 2>/dev/null |
+			sed -e "s:^$pkgbasedir/\.control\.::" -e "s:/\.provides-so::" |
+			tr -d '\n'
+		return 0
+	else
+		return 1
+	fi
+}
+
 subpkg_provides_so() {
-	grep -q -w "^so:$1" "$pkgbasedir"/.control.*/.provides-so 2>/dev/null
+	if grep -q -w "^so:$1" "$pkgbasedir"/.control.*/.provides-so 2>/dev/null; then
+		grep -l -H -w "^so:$1" "$pkgbasedir"/.control.*/.provides-so 2>/dev/null |
+			sed -e "s:^$pkgbasedir/\.control\.::" -e "s:/\.provides-so::" |
+			tr -d '\n'
+		return 0
+	else
+		return 1
+	fi
+}
+
+subpkg_provides_prefixed_pc() {
+	if [ -n "$pcprefix" ] && grep -q -w "^so:$pcprefix$1" \
+		"$pkgbasedir"/.control.*/.provides-pc 2>/dev/null; then
+		grep -l -H -w "^so:$pcprefix$1" "$pkgbasedir"/.control.*/.provides-pc 2>/dev/null |
+ 			sed -e "s:^$pkgbasedir/\.control\.::" -e "s:/\.provides-pc::" |
+			tr -d '\n'
+		return 0
+	else
+		return 1
+	fi
 }
 
 subpkg_provides_pc() {
-	grep -q -w "^pc:${1%%[<>=]*}" "$pkgbasedir"/.control.*/.provides-pc \
-		2>/dev/null
+	if grep -q -w "^${1%%[<>=]*}" "$pkgbasedir"/.control.*/.provides-pc \
+			2>/dev/null; then
+		grep -l -H -w "^${1%%[<>=]*}" "$pkgbasedir"/.control.*/.provides-pc |
+ 			sed -e "s:^$pkgbasedir/\.control\.::" -e "s:/\.provides-pc::" |
+			tr -d '\n'
+		return 0
+	else
+		return 1
+	fi
 }
 
 trace_apk_deps() {
@@ -1510,11 +1522,14 @@ trace_apk_deps() {
 		autodeps="$autodeps pkgconfig"
 	fi
 
-	for i in $(cat "$dir"/.needs-so 2>/dev/null); do
-		if subpkg_provides_so "$sonameprefix$i"; then
-			autodeps="$autodeps so:$sonameprefix$i"
-		elif $APK $apkroot info --quiet -e "so:$i"; then
-			autodeps="$autodeps so:$i"
+	[ -f "$dir"/.needs-so ] && for i in $(cat "$dir"/.needs-so); do
+		# first check if its provided by same apkbuild
+		grep -q -w "^$sonameprefix$i" "$dir"/.provides-so 2>/dev/null && continue
+
+		if subpkg_provides_prefixed_so "$i">/dev/null; then
+			warning "subpkg so dependency missing: $i in $(subpkg_provides_prefixed_so $i)"
+		elif subpkg_provides_so "$i">/dev/null; then
+			warning "subpkg so dependency missing: $i in $(subpkg_provides_so $i)"
 		else
 			missing="$missing $i"
 		fi
@@ -1526,39 +1541,56 @@ trace_apk_deps() {
 		so_files=$(find_so_files "$dir"/.rpaths $missing)
 		deppkgs=$($APK $apkroot info --quiet --who-owns $so_files)
 		for found in $deppkgs; do
-			if grep -q -x -F "depend = $found" "$dir"/.PKGINFO; then
-				warning "You can remove '$found' from depends"
-				continue
-			fi
-			autodeps="$autodeps $found"
+		    if ! grep -w "^depend = ${found}$" "$dir"/.PKGINFO >/dev/null ; then
+			    warning "so dependency missing: $found"
+		    fi
 		done
 	fi
 
+	# symlink targets
+	for i in $(sort -u "$dir"/.symlinks-needs 2>/dev/null); do
+		autodeps="$autodeps $i"
+	done
+
+	# python3 dependencies
+	for i in $(sort -u "$dir"/.python3-needs 2>/dev/null); do
+		autodeps="$autodeps $i"
+	done
+
+	# debuginfo conflict dependencies
+	local ver=$pkgver-r$pkgrel
+	for i in $(sort -u "$dir"/.debug-target-pkgs 2>/dev/null); do
+		autodeps="$autodeps !$i<$ver !$i>$ver"
+	done
+
 	# pkg-config depends
 	local provider
 	for i in $(sort -u "$dir"/.needs-pc 2>/dev/null); do
-		if subpkg_provides_pc "$pcprefix$i"; then
+		# first check if its provided by same apkbuild
+		grep -q -w "^$pcprefix$i" "$dir"/.provides-pc 2>/dev/null && continue
+
+		if subpkg_provides_prefixed_pc "$i" >/dev/null; then
 			autodeps="$autodeps pc:$pcprefix$i"
-		elif $APK $apkroot info --quiet -e "pc:$i"; then
-			provider=$($APK $apkroot search --from installed --quiet --exact "pc:${i%%[<>=~]*}")
-			if list_has "$provider" $depends_dev; then
-				warning "$provider should be removed from depends_dev"
+			warning "subpkg pkg-config depdency missing: $i in $(subpkg_provides_prefixed_pc $i)"
+		elif subpkg_provides_pc "$i"; then
+			provider=$($APK $apkroot search --quiet --exact "pc:$i")
+			if ! list_has "$provider" $depends_dev; then
+				warning "subpkg pkg-config dependency missing: $i in $(subpkg_provides_pc $i)"
 			fi
-			autodeps="$autodeps pc:$i"
 		else
-			warning "Could not find any provider for pc:$i"
+			warning "pkg-config dependency missing: $i"
 		fi
 	done
 
 	echo "# automatically detected:" >> "$dir"/.PKGINFO
-
-	touch "$dir"/.provides-dummy
-	sed 's/^/provides = /' "$dir"/.provides-* | sort -u >> "$dir"/.PKGINFO
-
-	for i in "$dir"/.needs-symlinks "$dir"/.needs-python3 "$dir"/.needs-debug; do
-		[ -f "$i" ] && cat "$i"
-	done | sed 's/^/depend = /' | sort -u >> "$dir"/.PKGINFO
-
+	if [ -f "$dir"/.provides-pc ]; then
+		sed 's/^/provides = pc:/' "$dir"/.provides-pc | sort -u \
+			>> "$dir"/.PKGINFO
+	fi
+	if [ -f "$dir"/.provides-py ]; then
+		sed 's/^/provides = /' "$dir"/.provides-py | sort -u \
+			>> "$dir"/.PKGINFO
+	fi
 	for i in $autodeps; do
 		echo "depend = $i"
 	done | sort -u >> "$dir"/.PKGINFO
@@ -2183,32 +2215,6 @@ fishcomp() {
 	default_fishcomp
 }
 
-default_pyc() {
-	depends=
-	pkgdesc="Precompiled Python bytecode for ${subpkgname%-pyc}"
-	install_if="${subpkgname%-pyc}=$pkgver-r$pkgrel pyc"
-
-	cd "$pkgdir"
-	local IFS=$'\n'
-	amove $(find usr/lib/python* -type d -name __pycache__)
-}
-
-pyc() {
-	default_pyc
-}
-
-default_nftrules() {
-	depends=
-	pkgdesc="$pkgdesc (nftables ruleset)"
-	install_if="nftables-rulesets ${subpkgname%-nftrules}=$pkgver-r$pkgrel"
-
-	amove 'usr/share/nftables.avail/*.nft' || :
-}
-
-nftrules() {
-	default_nftrules
-}
-
 is_function() {
 	type "$1" 2>&1 | head -n 1 | grep -E -q "is a (shell )?function"
 }
@@ -2488,8 +2494,7 @@ rootbld_qemu_arch() {
 	case "$CBUILD_ARCH" in
 	arm*) qarch="arm" ;;
 	x86)  qarch="i386" ;;
-	loongarch*) qarch="loongarch" ;;
-	*)    qarch="$CBUILD_ARCH" ;;
+	*)	qarch="$CBUILD_ARCH" ;;
 	esac
 
 	echo "${qarch}"
@@ -2931,8 +2936,8 @@ while getopts ":AcC:dD:fFhkKmnP:qrs:vV" opt; do
 	case $opt in
 		'A') echo "$CARCH"; exit 0;;
 		'c') enable_colors
-		     color_opt="-c";;
-		'C') cd "$OPTARG";;
+			 color_opt="-c";;
+		'C') chdir "$OPTARG";;
 		'd') nodeps="-d";;
 		'D') DESCRIPTION=$OPTARG;;
 		'f') force="-f";;
@@ -2941,7 +2946,7 @@ while getopts ":AcC:dD:fFhkKmnP:qrs:vV" opt; do
 		'k') keep="-k";;
 		'K') keep_build="-K";;
 		'm') disable_colors
-		     color_opt="-m";;
+			 color_opt="-m";;
 		'n') die "Use newapkbuild to create new aports";;
 		'P') REPODEST=$OPTARG;;
 		'q') quiet="-q";;
diff --git a/functions.sh.in b/functions.sh.in
index 994494e3a5..bad1f0732a 100644
--- a/functions.sh.in
+++ b/functions.sh.in
@@ -6,29 +6,8 @@ program=${0##*/}
 
 arch_to_triplet() {
 	case "$1" in
-	aarch64)	echo "aarch64-alpine-linux-musl" ;;
-	aarch64_be)	echo "aarch64_be-alpine-linux-musl" ;;
-	armel)		echo "armv5-alpine-linux-musleabi" ;;
-	armhf)		echo "armv6-alpine-linux-musleabihf" ;;
-	armv7)		echo "armv7-alpine-linux-musleabihf" ;;
-	loongarch32)	echo "loongarch32-alpine-linux-musl" ;;
-	loongarchx32)	echo "loongarchx32-alpine-linux-musl" ;;
-	loongarch64)	echo "loongarch64-alpine-linux-musl" ;;
-	mips)           echo "mips-alpine-linux-musl" ;;
-	mips64)         echo "mips64-alpine-linux-musl" ;;
-	mipsel)         echo "mipsel-alpine-linux-musl" ;;
-	mips64el)       echo "mips64el-alpine-linux-musl" ;;
-	ppc)		echo "powerpc-alpine-linux-musl" ;;
-	ppc64)		echo "powerpc64-alpine-linux-musl" ;;
-	ppc64le)	echo "powerpc64le-alpine-linux-musl" ;;
-	riscv32)	echo "riscv32-alpine-linux-musl" ;;
-	riscv64)	echo "riscv64-alpine-linux-musl" ;;
-	s390x)		echo "s390x-alpine-linux-musl" ;;
-	sh2eb)		echo "sh2eb-alpine-linux-muslfdpic" ;;
-	sh3)		echo "sh3-alpine-linux-musl" ;;
-	sh4)		echo "sh4-alpine-linux-musl" ;;
-	x86)		echo "i586-alpine-linux-musl" ;;
-	x86_64)		echo "x86_64-alpine-linux-musl" ;;
+	aarch64)	echo "aarch64-unknown-nto-qnx8.0.0" ;;
+	x86_64)		echo "x86_64-pc-nto-qnx8.0.0" ;;
 	*)		echo "unknown" ;;
 	esac
 }
@@ -40,28 +19,7 @@ arch_to_hostspec() {
 
 triplet_to_arch() {
 	case "$1" in
-	aarch64-*-*-*)		echo "aarch64" ;;
-	aarch64_be-*-*-*)	echo "aarch64_be" ;;
-	arm*-*-*-*eabi)		echo "armel" ;;
-	armv6*-*-*-*eabihf)	echo "armhf" ;;
-	armv7*-*-*-*eabihf)	echo "armv7" ;;
-	i[0-9]86-*-*-*)		echo "x86" ;;
-	loongarch32-*-*-*)	echo "loongarch32" ;;
-	loongarchx32-*-*-*)	echo "loongarchx32" ;;
-	loongarch64-*-*-*)	echo "loongarch64" ;;
-	mips-*-*-*)             echo "mips" ;;
-	mips64-*-*-*)           echo "mips64" ;;
-	mipsel-*-*-*)           echo "mipsel" ;;
-	mips64el-*-*-*)         echo "mips64el" ;;
-	powerpc-*-*-*)		echo "ppc" ;;
-	powerpc64-*-*-*)	echo "ppc64" ;;
-	powerpc64le-*-*-*)	echo "ppc64le" ;;
-	riscv32-*-*-*)		echo "riscv32" ;;
-	riscv64-*-*-*)		echo "riscv64" ;;
-	s390x-*-*-*)		echo "s390x" ;;
-	sh2eb-*-*-*)		echo "sh2eb" ;;
-	sh3-*-*-*)		echo "sh3" ;;
-	sh4-*-*-*)		echo "sh4" ;;
+	aarch64*-*-*-*)		echo "aarch64" ;;
 	x86_64-*-*-*)		echo "x86_64" ;;
 	*)
 		echo "Couldn't determine arch from triplet: $1" >&2
@@ -70,18 +28,6 @@ triplet_to_arch() {
 	esac
 }
 
-triplet_to_libc() {
-	case "$1" in
-	*-*-*-uclibc*)	echo "uclibc" ;;
-	*-*-*-musl*)	echo "musl" ;;
-	*-*-*-gnu*)	echo "glibc" ;;
-	*)
-		echo "Couldn't determine libc from triplet: $1" >&2
-		exit 1
-		;;
-	esac
-}
-
 readconfig() {
 	local _APORTSDIR _REPODEST _SRCDEST _ABUILD_USERDIR
 	local _CC=cc _CXX=c++ _LD=ld _CARCH _CHOST _CTARGET _CPPFLAGS _CFLAGS _CXXFLAGS _LDFLAGS
@@ -187,10 +133,8 @@ readconfig() {
 	fi
 
 	[ -z "$CARCH" ] && CARCH="$(triplet_to_arch $CHOST)"
-	[ -z "$CLIBC" ] && CLIBC="$(triplet_to_libc $CHOST)"
 	[ -z "$CBUILD_ARCH" ] && CBUILD_ARCH="$(triplet_to_arch $CBUILD)"
 	[ -z "$CTARGET_ARCH" ] && CTARGET_ARCH="$(triplet_to_arch $CTARGET)"
-	[ -z "$CTARGET_LIBC" ] && CTARGET_LIBC="$(triplet_to_libc $CTARGET)"
 
 	if [ "$CHOST" != "$CTARGET" ]; then
 		# setup environment for creating cross compiler

From f2428d020912f715424f70a9479d9901ef1882ec Mon Sep 17 00:00:00 2001
From: Leo Shen <i@szclsya.me>
Date: Thu, 30 Oct 2025 14:49:32 +0000
Subject: [PATCH 2/6] qnx: add BUILDINFO

- similar to Arch Linux's makepkg and pacman repro system
---
 abuild.in | 32 +++++++++++++++++++++++++++++++-
 1 file changed, 31 insertions(+), 1 deletion(-)

diff --git a/abuild.in b/abuild.in
index a5efb5e8b6..08e535d4a9 100644
--- a/abuild.in
+++ b/abuild.in
@@ -1089,6 +1089,7 @@ prepare_metafiles() {
 	[ -z "${name##* *}" ] && die "package name contains spaces"
 	local pkg="$name-$pkgver-r$pkgrel.apk"
 	local pkginfo="$controldir"/.PKGINFO
+	local buildinfo="$controldir"/.BUILDINFO
 	local i size
 
 	cd "${subpkgdir:-$pkgdir}"
@@ -1209,7 +1210,36 @@ prepare_metafiles() {
 		echo "install_if = $(echo $install_if)" >> "$pkginfo"
 	fi
 
-	local metafiles=".PKGINFO"
+        # QNX: BUILDINFO for dashboard and reproducible builds
+	cat >> "$buildinfo" <<-EOF
+		pkgname = $name
+		pkgver = $pkgver-r$pkgrel
+		pkgdesc = $pkgdesc
+		url = $url
+		builddate = $builddate
+		arch = ${subpkgarch:-$pkgarch}
+		origin = $pkgname
+		maintainer = $maintainer
+		packager = ${PACKAGER:-"Unknown"}
+		builddir = $builddir
+		startdir = $startdir
+	EOF
+	echo "apkbuild_sha512sum = $(sha512sum $APKBUILD | cut -d ' ' -f 1)" >> "$buildinfo"
+	# Use $url if vcs_url is not defined
+	echo "vcs_url = ${vcs_url:-$url}" >> "$buildinfo"
+	for i in $options; do
+		echo "option = $i" >> "$buildinfo"
+	done
+	local installed=$($APK list -I --manifest | sed 's: :=:g')
+	for i in $installed; do
+		local pkg=${i// /=}
+		case $pkg in
+			".makedepends-$pkgname="*) ;;
+			*) echo "installed = $pkg" >> "$buildinfo";;
+		esac
+	done
+
+	local metafiles=".PKGINFO .BUILDINFO"
 	for i in $install $triggers; do
 		local f=${i%=*}
 		local n=${f%.*}

From 07ee95b756124116c65fa130bd8e4533b76794b2 Mon Sep 17 00:00:00 2001
From: Leo Shen <i@szclsya.me>
Date: Tue, 21 Apr 2026 16:56:08 -0400
Subject: [PATCH 3/6] qnx: misc fixes

- fix unzip command arugment order
- add workaround for clang
---
 abuild.in    | 2 +-
 default.conf | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/abuild.in b/abuild.in
index 08e535d4a9..3bf8798567 100644
--- a/abuild.in
+++ b/abuild.in
@@ -542,7 +542,7 @@ default_unpack() {
 				tar -C "$srcdir" --zstd -xf "$s" || return 1;;
 			*.zip)
 				msg "Unpacking $s..."
-				unzip -n -q "$s" -d "$srcdir" || return 1;;
+				unzip -n -q -d "$srcdir" "$s" || return 1;;
 		esac
 	done
 }
diff --git a/default.conf b/default.conf
index 681a88f3c2..6f217eafdf 100644
--- a/default.conf
+++ b/default.conf
@@ -1,5 +1,5 @@
-export CFLAGS="-Os -fstack-clash-protection -Wformat -Werror=format-security"
-export CXXFLAGS="-Os -fstack-clash-protection -Wformat -Werror=format-security"
+export CFLAGS="-Os -fstack-clash-protection -Wformat -Werror=format-security -Qunused-arguments"
+export CXXFLAGS="-Os -fstack-clash-protection -Wformat -Werror=format-security -Qunused-arguments"
 export CXXFLAGS="$CXXFLAGS -D_GLIBCXX_ASSERTIONS=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_FAST"
 export LDFLAGS="-Wl,--as-needed,-O1,--sort-common"
 export GOFLAGS="-buildmode=pie -modcacherw -trimpath -buildvcs=false"

From c11f5a68f026f7f51d2c5025d6f4c0b3d1595a2f Mon Sep 17 00:00:00 2001
From: Aaron Bassett <abassett@qnx.com>
Date: Mon, 10 Aug 2026 04:19:08 +0000
Subject: [PATCH 4/6] fix symlink/python3 autodepends

---
 abuild.in | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/abuild.in b/abuild.in
index 3bf8798567..ad6874bb0d 100644
--- a/abuild.in
+++ b/abuild.in
@@ -1578,12 +1578,12 @@ trace_apk_deps() {
 	fi
 
 	# symlink targets
-	for i in $(sort -u "$dir"/.symlinks-needs 2>/dev/null); do
+	for i in $(sort -u "$dir"/.needs-symlinks 2>/dev/null); do
 		autodeps="$autodeps $i"
 	done
 
 	# python3 dependencies
-	for i in $(sort -u "$dir"/.python3-needs 2>/dev/null); do
+	for i in $(sort -u "$dir"/.needs-python3 2>/dev/null); do
 		autodeps="$autodeps $i"
 	done
 

From 6a2894ae2fbe14a567b7953d545e53e69c8481ab Mon Sep 17 00:00:00 2001
From: jscaff <jscaffidi@qnx.com>
Date: Thu, 10 Sep 2026 14:14:50 +0000
Subject: [PATCH 5/6] fix ub

---
 abuild-gzsplit.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/abuild-gzsplit.c b/abuild-gzsplit.c
index ac55c54a6e..64f477e3f6 100644
--- a/abuild-gzsplit.c
+++ b/abuild-gzsplit.c
@@ -20,7 +20,13 @@ static int find_section(const char *buf, size_t bufsize, const char *str) {
 int main(void)
 {
 	char obuf[8*1024], ibuf[8*1024];
-	z_stream zs;
+	z_stream zs = {
+		.next_in = 0,
+		.avail_in = 0,
+		.zalloc = Z_NULL,
+		.zfree = Z_NULL,
+		.opaque = Z_NULL,
+	};
 	int r = Z_OK, rc = 1, fd = -1;
 	size_t len;
 

From 9e8151a4ebb4f3d30f5cda7be728bc8636bd080e Mon Sep 17 00:00:00 2001
From: jscaff <jscaffidi@qnx.com>
Date: Tue, 29 Sep 2026 13:12:07 +0000
Subject: [PATCH 6/6] fix pkginfo provides having double prefixes

Likely poorly fixed merge conflict from alpine/ca9e4db4a1cca4fba82a84fad7020908b7dd1c07
---
 abuild.in | 11 +++--------
 1 file changed, 3 insertions(+), 8 deletions(-)

diff --git a/abuild.in b/abuild.in
index ad6874bb0d..5f01965b5b 100644
--- a/abuild.in
+++ b/abuild.in
@@ -1613,14 +1613,9 @@ trace_apk_deps() {
 	done
 
 	echo "# automatically detected:" >> "$dir"/.PKGINFO
-	if [ -f "$dir"/.provides-pc ]; then
-		sed 's/^/provides = pc:/' "$dir"/.provides-pc | sort -u \
-			>> "$dir"/.PKGINFO
-	fi
-	if [ -f "$dir"/.provides-py ]; then
-		sed 's/^/provides = /' "$dir"/.provides-py | sort -u \
-			>> "$dir"/.PKGINFO
-	fi
+	touch "$dir"/.provides-dummy
+	sed 's/^/provides = /' "$dir"/.provides-* | sort -u >> "$dir"/.PKGINFO
+
 	for i in $autodeps; do
 		echo "depend = $i"
 	done | sort -u >> "$dir"/.PKGINFO
